execution.bs

```diff
@@ -3173,67 +3196,91 @@ Insert this section as a new subclause between Header `<stop_token>` synopsis <b
 <ins>
 <blockquote>
-1. The `stoppable_token` concept checks for the basic interface of a “stop token” which is copyable and allows polling to see if stop has been requested and also whether a stop request is possible. It also requires an associated nested template-type-alias,
-    `T::callback_type<CB>`, that identifies the stop-callback type to use to register a callback to be executed if a stop-request is ever made on a stoppable_token of type, `T`. The `stoppable_token_for` concept checks for a stop token type compatible with a given
-    callback type. The `unstoppable_token` concept checks for a stop token type that does not allow stopping.
+1. The `stoppable_token` concept checks for the basic interface of a stop token
+    that is copyable and allows polling to see if stop has been requested and
+    also whether a stop request is possible. For a stop token type `T` and a type
+    `CB` that is callable with no arguments, the type `T::callback_type<CB>` shall
+    be well-formed and name the stop callback type to use to register a callback
+    to be executed if a stop request is ever made on a `stoppable_token` of type
+    `T`. The `stoppable_token_for` concept checks for a stop token type compatible
+    with a given callback type. The `unstoppable_token` concept checks for a stop
+    token type that does not allow stopping.
 
 <pre>
-template&lt;typename T>
+template&lt;class T>
   concept stoppable_token =
-    copy_constructible&lt;T> &&
-    move_constructible&lt;T> &&
-    is_nothrow_copy_constructible_v&lt;T> &&
-    is_nothrow_move_constructible_v&lt;T> &&
+    copyable&lt;T> &&
     equality_comparable&lt;T> &&
-    requires (const T& token) {
-      { token.stop_requested() } noexcept -> boolean-testable;
-      { token.stop_possible() } noexcept -> boolean-testable;
+    requires (const T t) {
+      { T{t} } noexcept;
+      { t.stop_requested() } noexcept -> <i>boolean-testable</i>;
+      { t.stop_possible() } noexcept -> <i>boolean-testable</i>;
       typename <i>check-type-alias-exists</i>&lt;T::template callback_type>;
     };
 
-template&lt;typename T, typename CB, typename Initializer = CB>
+template&lt;class T, class CB, class Initializer = CB>
   concept stoppable_token_for =
     stoppable_token&lt;T> &&
     invocable&lt;CB> &&
-    requires {
-      typename T::template callback_type&lt;CB>;
-    } &&
     constructible_from&lt;CB, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, T, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, T&, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, const T, Initializer> &&
-    constructible_from&lt;typename T::template callback_type&lt;CB>, const T&, Initializer>;
+    requires (const T t) {
+      typename stop_callback_for_t&lt;T, CB>;
+      []&lt;class T>(T&& t)
+        requires constructible_from&lt;stop_callback_for_t&lt;T, CB>, U, Initializer>
+      {}(t);
```

Comment:
I'm not convinced this is better than spamming out four almost identical `constructible_from` constraints.

Reply:
I preffer previous specification. We could also move the non-const& stop_token to semantics requirements. So impls will not need to check it.

Reply:
And leave a note regarding implicit expression variance.

Reply:
Another way to skin the cat would be to generalize implicit expression variations to also include nested requirements. Then this could be:

```
template<class T, class CB, class Initializer = CB>
  concept stoppable_token_for =
    stoppable_token<T> &&
    invocable<CB> &&
    requires (const T t) {
      typename stop_callback_for_t<T, CB>;
      requires constructible_from<stop_callback_for_t<T, CB>, decltype(t), Initializer>;
    };
```

I kinda like that best. Thoughts, @tomaszkam ?

Reply:
I will just add point 4 for the list pf requirements for `stoppable_token_for`:
```
  constructible_from&lt;typename T::template callback_type&lt;CB>, T, Initializer>,  
  constructible_from&lt;typename T::template callback_type&lt;CB>, T&, Initializer>,  
  constructible_from&lt;typename T::template callback_type&lt;CB>, const T, Initializer> are modeled.
```
Or put them at begining.